homeassistant/components/osoenergy/sensor.py

```diff
@@ -0,0 +1,179 @@
+"""Support for OSO Energy sensors."""
+
+from collections.abc import Callable
+from dataclasses import dataclass
+
+from apyosoenergyapi import OSOEnergy
+from apyosoenergyapi.helper.const import OSOEnergySensorData
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import UnitOfEnergy, UnitOfPower, UnitOfVolume
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.typing import StateType
+import homeassistant.util.dt as dt_util
+
+from . import OSOEnergyEntity
+from .const import DOMAIN
+
+
+def _get_local_hour(utc_hour: int):
+    """Get the local hour."""
+    now = dt_util.utcnow()
+    now_local = dt_util.now()
+    utc_time = now.replace(hour=utc_hour, minute=0, second=0, microsecond=0)
+    local_hour = dt_util.as_local(utc_time)
+    return local_hour.replace(
+        year=now_local.year, month=now_local.month, day=now_local.day
+    )
+
+
+def _convert_profile_to_local(values):
+    """Convert UTC profile to local."""
+    profile = [None] * 24
+    for hour in range(24):
+        local_hour = _get_local_hour(hour)
+        profile[local_hour.hour] = values[hour]
+
+    return profile
+
+
+@dataclass(frozen=True, kw_only=True)
+class OSOEnergySensorEntityDescription(SensorEntityDescription):
+    """Class describing OSO Energy heater sensor entities."""
+
+    value: Callable[[OSOEnergy], StateType]
+
+
+SENSOR_TYPES: dict[str, OSOEnergySensorEntityDescription] = {
+    "profile": OSOEnergySensorEntityDescription(
+        key="profile",
+        translation_key="profile",
+        value=lambda device: _convert_profile_to_local(device.state),
+    ),
+    "heater_mode": OSOEnergySensorEntityDescription(
+        key="heater_mode",
+        translation_key="heater_mode",
+        device_class=SensorDeviceClass.ENUM,
+        options=[
+            "auto",
+            "manual",
+            "off",
+            "legionella",
+            "powersave",
+            "extraenergy",
+            "voltage",
+            "ffr",
+        ],
+        value=lambda device: device.state.lower(),
```

Comment:
I don't think `device` is a great name here, all the sensors provide different measurements from the same boiler, right?
I'd suggest to call it `sensor_data` or `sensor_state` or something like that. If you want to use the same name also in the water heater entity, maybe name it `endpoint_data` or `endpoint_state`.